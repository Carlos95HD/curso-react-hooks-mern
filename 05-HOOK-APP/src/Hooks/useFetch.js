import { useState, useEffect, useRef } from 'react'

export const useFetch = (url) => {

  const estaMontado = useRef(true)
  const [state, setstate] = useState({data: null, loading:true, error: null})

  useEffect(() => {

    return () => {
      estaMontado.current = false;
    }
  }, [])

  useEffect(() => {

    setstate({data: null, loading:true, error: null})

    fetch(url)
    .then(resp => resp.json() )
    .then( data => {

      if (estaMontado.current) {
        setstate({
          loading: false,
          error: null,
          data
        })
      }

    }).catch(() => {
      setstate({
        data: null,
        loading:false, 
        error: 'No se pudo cargar la informaciÃ³n'
      })
    })

  }, [url])

  return state;
}

