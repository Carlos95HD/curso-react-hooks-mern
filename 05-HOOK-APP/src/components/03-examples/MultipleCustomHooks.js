import React from 'react'
import { useCounter } from '../../Hooks/useCounter'
import { useFetch } from '../../Hooks/useFetch'
import '../02-useEffect/effects.css'

export const MultipleCustomHooks = () => {

  const {counter, increment} = useCounter(1)

  const { loading, data } = useFetch(`https://breakingbadapi.com/api/quotes/${counter}`)
  //si !!data devuelve false no retornara data[0] 
  const { author, quote } = !!data && data[0]


  return (
    <div>
      <h1>BreakingBad Quotes</h1>
      <hr />

      {
        loading ?
        <div className='alert alert-info text-center'>
          loading...
        </div>
        :
        <blockquote className='blockquote text-right'>
          <p className="mb-0">{ quote }</p>
          <footer className='blockquote-footer'>{ author }</footer>
        </blockquote>
      }

      <button 
      className='btn btn-primary'
      onClick={ increment }
      >Siguiente
      </button>

    </div>
  )
}
