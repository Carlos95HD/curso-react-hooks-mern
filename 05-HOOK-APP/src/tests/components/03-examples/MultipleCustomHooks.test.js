import React from 'react'
import { shallow } from 'enzyme'
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks'
import { useFetch } from '../../../Hooks/useFetch'
import { useCounter } from '../../../Hooks/useCounter'
jest.mock('../../../Hooks/useFetch')
jest.mock('../../../Hooks/useCounter')

describe('Pruebas en MultiplesCustomhooks', () => {

  test('should mostrarse correctamente', () => {

    useCounter.mockReturnValue({
      counter: 10,
      increment: () => {}
    })

    useFetch.mockReturnValue({
      data: null,
      loading:true,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect( wrapper ).toMatchSnapshot();
  });
  
  test('should mostrar informaciÃ³n', () => {

    useFetch.mockReturnValue({
      data: [{
        author:'Carlos',
        quote:'Hola Mundo'
      }],
      loading:false,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find('.alert').exists() ).toBe(false)
    expect(wrapper.find('.mb-0').text().trim() ).toBe('Hola Mundo')
    expect(wrapper.find('footer').text().trim() ).toBe('Carlos')

  })

  
})
