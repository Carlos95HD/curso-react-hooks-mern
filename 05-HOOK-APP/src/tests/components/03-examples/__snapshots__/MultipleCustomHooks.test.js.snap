// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultiplesCustomhooks should mostrarse correctamente 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente
  </button>
</div>
`;
